File: smart_copy/CopyEngine.ts
import * as fs from 'fs';
import * as path from 'path';
import { Transform } from 'stream';

export enum CopyStatus {
  PENDING = 'pending',
  IN_PROGRESS = 'in_progress',
  COMPLETED = 'completed',
  FAILED = 'failed',
  CANCELLED = 'cancelled'
}

export interface CopyFile {
  id: string;
  source: string;
  destination: string;
  status: CopyStatus;
  progress: number;
  speed: number;
  eta: number;
  size: number;
  copiedBytes: number;
}

export interface GlobalCopyStats {
  totalBytesCopied: number;
  elapsedSeconds: number;
  averageSpeed: number;
  filesCompleted: number;
  totalFiles: number;
}

export class CopyEngine {
  private queue: Map<string, CopyFile> = new Map();
  private isCancelled = false;
  private totalBytesCopied = 0;
  private startTime = 0;
  private readonly bufferSize: number;
  private readonly threads: number;

  constructor(bufferSize: number = 8 * 1024 * 1024, threads: number = 2) {
    this.bufferSize = bufferSize;
    this.threads = threads;
  }

  addFile(id: string, source: string, destination: string, size: number): CopyFile {
    const copyFile: CopyFile = {
      id,
      source,
      destination,
      status: CopyStatus.PENDING,
      progress: 0,
      speed: 0,
      eta: 0,
      size,
      copiedBytes: 0
    };
    this.queue.set(id, copyFile);
    return copyFile;
  }

  async start(onProgress: (file: CopyFile, global: GlobalCopyStats) => void): Promise<void> {
    this.isCancelled = false;
    this.startTime = Date.now();

    const files = Array.from(this.queue.values()).filter(f => f.status === CopyStatus.PENDING);

    for (const file of files) {
      if (this.isCancelled) break;
      await this.copyFile(file, onProgress);
    }
  }

  private async copyFile(file: CopyFile, onProgress: (file: CopyFile, global: GlobalCopyStats) => void): Promise<void> {
    return new Promise((resolve, reject) => {
      try {
        file.status = CopyStatus.IN_PROGRESS;
        const startMs = Date.now();
        let copiedBytes = 0;

        const input = fs.createReadStream(file.source, { highWaterMark: this.bufferSize });
        const output = fs.createWriteStream(file.destination);

        input.on('data', (chunk: Buffer) => {
          copiedBytes += chunk.length;
          this.totalBytesCopied += chunk.length;
          file.copiedBytes = copiedBytes;

          const elapsed = (Date.now() - startMs) / 1000;
          const speedMBps = (copiedBytes / 1024 / 1024) / elapsed;
          const remaining = file.size - copiedBytes;
          const eta = speedMBps > 0 ? (remaining / 1024 / 1024) / speedMBps : 0;

          file.progress = (copiedBytes / file.size) * 100;
          file.speed = speedMBps;
          file.eta = Math.ceil(eta);

          const global = this.getGlobalStats();
          onProgress(file, global);
        });

        input.on('error', (err) => {
          file.status = CopyStatus.FAILED;
          reject(err);
        });

        output.on('error', (err) => {
          file.status = CopyStatus.FAILED;
          reject(err);
        });

        output.on('finish', () => {
          file.status = CopyStatus.COMPLETED;
          file.progress = 100;
          const global = this.getGlobalStats();
          onProgress(file, global);
          resolve();
        });

        input.pipe(output);
      } catch (error) {
        file.status = CopyStatus.FAILED;
        reject(error);
      }
    });
  }

  cancel(): void {
    this.isCancelled = true;
  }

  getGlobalStats(): GlobalCopyStats {
    const elapsed = Math.max(1, (Date.now() - this.startTime) / 1000);
    const averageSpeed = (this.totalBytesCopied / 1024 / 1024) / elapsed;
    const completed = Array.from(this.queue.values()).filter(f => f.status === CopyStatus.COMPLETED).length;

    return {
      totalBytesCopied: this.totalBytesCopied,
      elapsedSeconds: Math.ceil(elapsed),
      averageSpeed: averageSpeed,
      filesCompleted: completed,
      totalFiles: this.queue.size
    };
  }
}

File: smart_copy/CopyEngine.kt
package com.smartcopy.app

import java.io.File
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import kotlin.math.min

data class CopyFile(
    val source: File,
    val destination: File,
    var status: CopyStatus = CopyStatus.PENDING,
    var progress: Float = 0f,
    var speed: Double = 0.0,
    var eta: Long = 0
)

enum class CopyStatus {
    PENDING, IN_PROGRESS, COMPLETED, FAILED, CANCELLED
}

class CopyEngine(private val metrics: PerformanceMetrics) {
    private val executor: ExecutorService = Executors.newFixedThreadPool(metrics.recommendedThreads)
    private var totalBytesCopied = 0L
    private var startTime = 0L
    private var isCancelled = false
    
    fun addFiles(files: List<CopyFile>) {
        files.forEach { file ->
            executor.submit { copyFileWithOptimization(file) }
        }
    }
    
    fun cancelAll() {
        isCancelled = true
        executor.shutdownNow()
    }
    
    private fun copyFileWithOptimization(copyFile: CopyFile) {
        if (isCancelled) return
        
        try {
            copyFile.status = CopyStatus.IN_PROGRESS
            startTime = System.currentTimeMillis()
            
            val source = copyFile.source
            val destination = copyFile.destination
            val totalSize = source.length()
            val bufferSize = metrics.optimalBufferSize
            
            source.inputStream().use { input ->
                destination.outputStream().use { output ->
                    val buffer = ByteArray(bufferSize)
                    var bytesRead: Int
                    var copiedBytes = 0L
                    val startMilis = System.currentTimeMillis()
                    
                    while (input.read(buffer).also { bytesRead = it } > 0 && !isCancelled) {
                        output.write(buffer, 0, bytesRead)
                        copiedBytes += bytesRead
                        totalBytesCopied += bytesRead
                        
                        val elapsed = (System.currentTimeMillis() - startMilis) / 1000.0
                        val speedMBps = (copiedBytes / 1024.0 / 1024.0) / elapsed
                        val remainingBytes = totalSize - copiedBytes
                        val etaSeconds = if (speedMBps > 0) remainingBytes / 1024.0 / 1024.0 / speedMBps else 0.0
                        
                        copyFile.progress = (copiedBytes.toFloat() / totalSize) * 100
                        copyFile.speed = speedMBps
                        copyFile.eta = etaSeconds.toLong()
                    }
                    
                    if (!isCancelled) {
                        copyFile.status = CopyStatus.COMPLETED
                    }
                }
            }
        } catch (e: Exception) {
            copyFile.status = CopyStatus.FAILED
        }
    }
    
    fun getGlobalStats(): GlobalCopyStats {
        val elapsed = (System.currentTimeMillis() - startTime) / 1000.0
        val globalSpeed = if (elapsed > 0) (totalBytesCopied / 1024.0 / 1024.0) / elapsed else 0.0
        
        return GlobalCopyStats(
            totalBytesCopied = totalBytesCopied,
            elapsedSeconds = elapsed.toLong(),
            averageSpeed = globalSpeed
        )
    }
}

data class GlobalCopyStats(
    val totalBytesCopied: Long,
    val elapsedSeconds: Long,
    val averageSpeed: Double
)
